subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java

```diff
@@ -57,12 +63,21 @@ <T> Provider<T> toolFor(JavaToolchainSpec spec, Transformer<T, JavaToolchain> to
     }
 
     Provider<JavaToolchain> findMatchingToolchain(JavaToolchainSpec filter) {
+        JavaToolchainSpecInternal filterInternal = (JavaToolchainSpecInternal) filter;
+        if (!filterInternal.isValid()) {
+            DeprecationLogger.deprecate("Using toolchain specifications without setting a language version")
+                .withAdvice("Consider configuring the language version.")
```

Comment:
I think the message should have a documentation link. And documentation should be edited to indicate what is a valid toolchain spec and what is an invalid one.

Reply:
Done